### PR TITLE
cpp_scopet::lookup methods now return an id set

### PR DESCRIPTION
**Pull request by kroening**
- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [X] My contribution is formatted in line with CODING_STANDARD.md.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [X] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: bfe7eed08ffc75ed50532fe1210a0bdfb5cb7983).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86364498

File: src/cpp/cpp_instantiate_template.cpp [hunk 1]
```diff
@@ -335,8 +335,8 @@ const symbolt &cpp_typecheckt::instantiate_template(
   {
     cpp_scopet &scope=cpp_scopes.get_scope(subscope_name);
 
-    cpp_scopet::id_sett id_set;
-    scope.lookup(template_symbol.base_name, cpp_scopet::SCOPE_ONLY, id_set);
+    cpp_scopet::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_instantiate_template.cpp]: `const`

File: src/cpp/cpp_declarator_converter.cpp [hunk 2]
```diff
@@ -496,10 +494,9 @@ symbolt &cpp_declarator_convertert::convert_new_symbol(
 
   if(!is_code)
   {
-    cpp_scopest::id_sett id_set;
-
-    cpp_typecheck.cpp_scopes.current_scope().lookup(
-      base_name, cpp_scopet::SCOPE_ONLY, id_set);
+    cpp_scopest::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_declarator_converter.cpp]: `const`

File: src/cpp/cpp_declarator_converter.cpp [hunk 3]
```diff
@@ -207,10 +207,8 @@ symbolt &cpp_declarator_convertert::convert(
 
     if(symbol.type.id()=="cpp-template-type")
     {
-      cpp_scopet::id_sett id_set;
-
-      scope->lookup_identifier(
-        symbol.name, cpp_idt::id_classt::TEMPLATE_PARAMETER, id_set);
+      cpp_scopet::id_sett id_set = scope->lookup_identifier(
```
Comment (tautschnig) [inline src/cpp/cpp_declarator_converter.cpp]: `const`

File: src/cpp/cpp_scope.cpp [hunk 4]
```diff
@@ -208,10 +212,8 @@ cpp_scopet &cpp_scopet::new_scope(const irep_idt &new_scope_name)
   return (cpp_scopet &)id;
 }
 
-
 bool cpp_scopet::contains(const irep_idt &base_name)
 {
-  id_sett id_set;
-  lookup(base_name, SCOPE_ONLY, id_set);
+  const id_sett id_set = lookup(base_name, SCOPE_ONLY);
```
Comment (tautschnig) [inline src/cpp/cpp_scope.cpp]: Maybe this should just be rewritten as `return !lookup(base_name, SCOPE_ONLY).empty();`

File: src/cpp/cpp_typecheck_compound_type.cpp [hunk 5]
```diff
@@ -105,8 +105,8 @@ cpp_scopet &cpp_typecheckt::tag_scope(
   // Check if we have it already. If so, take it.
 
   // we should only look for tags, but we don't
-  cpp_scopet::id_sett id_set;
-  cpp_scopes.current_scope().lookup(base_name, cpp_scopet::RECURSIVE, id_set);
+  cpp_scopet::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_compound_type.cpp]: `const`

File: src/cpp/cpp_typecheck_compound_type.cpp [hunk 6]
```diff
@@ -816,10 +816,8 @@ void cpp_typecheckt::put_compound_into_scope(
   else
   {
     // check if it's already there
-    cpp_scopest::id_sett id_set;
-
-    cpp_scopes.current_scope().lookup(
-      base_name, cpp_scopet::SCOPE_ONLY, id_set);
+    cpp_scopest::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_compound_type.cpp]: `const`

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 7]
```diff
@@ -914,15 +914,12 @@ cpp_scopet &cpp_typecheck_resolvet::resolve_scope(
       template_args=to_cpp_template_args_non_tc(*pos);
     else if(pos->id()=="::")
     {
-      cpp_scopest::id_sett id_set;
-
       if(template_args.is_not_nil())
       {
-        cpp_typecheck.cpp_scopes.current_scope().lookup(
+        auto id_set = cpp_typecheck.cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_resolve.cpp]: `const`

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 8]
```diff
@@ -946,10 +943,9 @@ cpp_scopet &cpp_typecheck_resolvet::resolve_scope(
       }
       else
       {
-        cpp_typecheck.cpp_scopes.current_scope().lookup(
+        auto id_set = cpp_typecheck.cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_resolve.cpp:946]: `const`

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 9]
```diff
@@ -1259,12 +1255,8 @@ cpp_scopet &cpp_typecheck_resolvet::resolve_namespace(
   const source_locationt &source_location=cpp_name.source_location();
   bool qualified=cpp_name.is_qualified();
 
-  cpp_scopest::id_sett id_set;
-
-  cpp_typecheck.cpp_scopes.current_scope().lookup(
-    base_name,
-    qualified?cpp_scopet::QUALIFIED:cpp_scopet::RECURSIVE,
-    id_set);
+  cpp_scopest::id_sett id_set = cpp_typecheck.cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_resolve.cpp]: `const`

File: src/cpp/cpp_typecheck_resolve.cpp [hunk 10]
```diff
@@ -1765,9 +1757,9 @@ void cpp_typecheck_resolvet::guess_template_args(
       irep_idt base_name;
       resolve_scope(cpp_name, base_name, template_args);
 
-      cpp_scopest::id_sett id_set;
-      cpp_typecheck.cpp_scopes.current_scope().lookup(
-        base_name, cpp_scopet::RECURSIVE, id_set);
+      cpp_scopest::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_resolve.cpp]: `const`

File: src/cpp/cpp_typecheck_template.cpp [hunk 11]
```diff
@@ -83,12 +83,11 @@ void cpp_typecheckt::typecheck_class_template(
   // Check if the name is already used by a different template
   // in the same scope.
   {
-    cpp_scopet::id_sett id_set;
-    cpp_scopes.current_scope().lookup(
-      base_name,
-      cpp_scopet::SCOPE_ONLY,
-      cpp_scopet::TEMPLATE,
-      id_set);
+    cpp_scopet::id_sett id_set=
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_template.cpp]: `const`

File: src/cpp/cpp_typecheck_template.cpp [hunk 12]
```diff
@@ -345,13 +344,10 @@ void cpp_typecheckt::typecheck_class_template_member(
   }
 
   // let's find the class template this function template belongs to.
-  cpp_scopet::id_sett id_set;
-
-  cpp_scopes.current_scope().lookup(
+  cpp_scopet::id_sett id_set = cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_template.cpp]: `const`

File: src/cpp/cpp_typecheck_template.cpp [hunk 13]
```diff
@@ -539,9 +535,8 @@ void cpp_typecheckt::convert_class_template_specialization(
 
   // get the template symbol
 
-  cpp_scopest::id_sett id_set;
-  cpp_scopes.current_scope().lookup(
-    base_name, cpp_scopet::SCOPE_ONLY, cpp_idt::id_classt::TEMPLATE, id_set);
+  cpp_scopest::id_sett id_set = cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_template.cpp]: `const`

File: src/cpp/cpp_typecheck_template.cpp [hunk 14]
```diff
@@ -656,9 +651,8 @@ void cpp_typecheckt::convert_template_function_or_member_specialization(
     std::string base_name=
       cpp_name.get_sub()[0].get(ID_identifier).c_str();
 
-    cpp_scopest::id_sett id_set;
-    cpp_scopes.current_scope().lookup(
-      base_name, cpp_scopet::SCOPE_ONLY, id_set);
+    cpp_scopest::id_sett id_set =
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_template.cpp]: `const`

File: src/cpp/cpp_typecheck_using.cpp [hunk 15]
```diff
@@ -27,10 +27,9 @@ void cpp_typecheckt::convert(cpp_usingt &cpp_using)
   resolver.resolve_scope(cpp_using.name(), base_name, template_args);
 
   bool qualified=cpp_using.name().is_qualified();
-  cpp_scopest::id_sett id_set;
 
-  cpp_scopes.current_scope().lookup(
-    base_name, qualified?cpp_scopet::QUALIFIED:cpp_scopet::RECURSIVE, id_set);
+  cpp_scopest::id_sett id_set = cpp_scopes.current_scope().lookup(
```
Comment (tautschnig) [inline src/cpp/cpp_typecheck_using.cpp]: `const`

**Review by tautschnig — APPROVED** [review]
I may be wrong on some of the below, but it looks like a lot of const-ness could be added.

File: src/cpp/cpp_scope.cpp (hunk 4 above)
Comment (kroening) [reply src/cpp/cpp_scope.cpp]: ok

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 89b2bb37f3ef807c0dffbea1ba63a614b96f0232).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/86369058

**Pull request merged**